libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileUtils.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.runtime.policy;
+
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.io.File;
+import java.util.Comparator;
+
+import static java.lang.Character.isLetter;
+
+public class FileUtils {
+
+    private FileUtils() {}
+
+    /**
+     * For our lexicographic sort trick to work correctly, we must have path separators sort before
+     * any other character so that files in a directory appear immediately after that directory.
+     * For example, we require [/a, /a/b, /a.xml] rather than the natural order [/a, /a.xml, /a/b].
+     */
+    static final Comparator<String> PATH_ORDER = (s1, s2) -> {
+        int len1 = s1.length();
+        int len2 = s2.length();
+        int lim = Math.min(len1, len2);
+        for (int k = 0; k < lim; k++) {
+            char c1 = s1.charAt(k);
+            char c2 = s2.charAt(k);
```

Comment:
Using `charAt` tends to cause problems for characters outside the basic multilingual plane (like emojis), but I suspect it doesn't matter in this case because we're only comparing for order, not trying to identify the actual characters.

As silly as it sounds, it might be good to have at least one test with an emoji to demonstrate that this logic works with characters outside the BMP.

Reply:
I added a test and there were no surprises.
I think we are OK here: as you said, as long as the order is consistent we don't care much. One important caveat is that we do not match the "special" char (/ or \) with a random surrogate by mistake, but surrogates are in a whole different range, so that won't be an issue.